arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp

```diff
@@ -0,0 +1,1264 @@
+  if (params.length() != 1) {
+    return EvalError("expected a single argument");
+  }
+  result.add(VPackValue(ValueConsideredFalse(params.at(0))));
+  return {};
+}
+
+EvalResult Prim_PrintLn(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  ctx.print(paramsToString(params));
+  result.add(VPackSlice::noneSlice());
+  return {};
+}
+
+void Machine::print(const std::string& msg) const {
+  if (printCallback) {
+    printCallback(msg);
+  } else {
+    std::cerr << msg << std::endl;
+  }
+}
+
+EvalResult Prim_Error(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  return EvalError(paramsToString(params));
+}
+
+EvalResult Prim_List(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  VPackArrayBuilder ab(&result);
+  result.add(VPackArrayIterator(params));
+  return {};
+}
+
+EvalResult checkArrayParams(VPackSlice const& arr, VPackSlice const& index) {
+  if (!arr.isArray()) {
+    return EvalError("expect first parameter to be an array");
+  }
+
+  if (!index.isNumber()) {
+    return EvalError("expect second parameter to be a number");
+  }
+
+  if (index.getInt() < 0) {
+    return EvalError("number cannot be less than zero");
+  }
+
+  if (index.getUInt() > (arr.length() - 1)) {
+    return EvalError("array index is out of bounds");
+  }
+
+  return {};
+}
+
+EvalResult Prim_ListRef(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (!params.isArray() && params.length() != 2) {
+    return EvalError("expected exactly two parameters");
+  }
+
+  auto&& arr = params.at(0);
+  auto&& index = params.at(1);
+
+  auto check = checkArrayParams(arr, index);
+  if (check.fail()) {
+    return check;
+  }
+
+  result.add(arr.at(index.getUInt()));
+
+  return {};
+}
+
+EvalResult Prim_ListSet(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (!params.isArray() && params.length() != 3) {
+    return EvalError("expected exactly two parameters");
+  }
+
+  auto&& arr = params.at(0);
+  auto&& index = params.at(1);
+  auto&& value = params.at(2);
+
+  auto check = checkArrayParams(arr, index);
+  if (check.fail()) {
+    return check;
+  }
+
+  uint64_t pos = 0;
+  result.openArray();
+  for (auto&& element : VPackArrayIterator(arr)) {
+    if (pos == index.getUInt()) {
+      result.add(value);
+    } else {
+      result.add(element);
+    }
+
+    pos++;
+  }
+  result.close();
+
+  return {};
+}
+
+namespace {
+EvalResultT<VPackSlice> ReadAttribute(VPackSlice slice, VPackSlice key) {
+  if (!slice.isObject()) {
+    return EvalError("expect first parameter to be an object");
+  }
+
+  if (key.isString()) {
+    return slice.get(key.stringRef());
+  } else if (key.isArray()) {
+    std::vector<VPackStringRef> path;
+    for (auto&& pathStep : VPackArrayIterator(key)) {
+      if (!pathStep.isString()) {
+        return EvalError("expected string in key arrays");
+      }
+      path.emplace_back(pathStep.stringRef());
+    }
+
+    return slice.get(path);
```

Comment:
Instead of building a vector first, and then calling `get()` with the vector, it would be more efficient to make the lookups one by one, and update `slice` after each. This saves one iteration over the array and all memory allocations here.